<a href="https://colab.research.google.com/github/neto-riga/metodos_numericos_2/blob/main/15_extrapolacion_de_romberg.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## [Regresar al ínidce](https://github.com/neto-riga/metodos_numericos_2/blob/main/portafolio.md#índice)

# Ejercicio 15. Extrapolación de Romberg
---

Estimar el valor de la siguiente integral 

$$
\int_{0.5}^{2} x^{3} \frac{\sqrt{3+2x^{2}}}{4} dx 
$$

1. Empleanto la extrapolación de Romber, iniciando con $h=$0.1, 0.05, 0.025.
2. Aplicar la extrapolación de Romberg.
3. Hasta alcanzar una precisión de 6 cifras.


In [ ]:
import numpy as np 
import pandas as pd

Implementamos el siguiente código para llenar listas

In [ ]:
def Listas(val,h,cant):
  f = lambda x: (x**3) * ((np.sqrt(3 + (2*(x**2))))/(4))
  x=[val]
  fx = [f(x[0])]
  for i in range(cant):
    x.append(x[i]+h)
    fx.append(f(x[i+1]))
  return x,fx

Creamos y tabulamos para $x$ y $f(x)$ con h=0.1

In [ ]:
x,fx = Listas(0.5,0.1,15)
tabla = {'x': x,'f(x)': fx}
df = pd.DataFrame(tabla)
df.head(5)

x      f(x)
0  0.5  0.058463
1  0.6  0.104151
2  0.7  0.171071
3  0.8  0.264808
4  0.9  0.391732

Llenamos las listas y tabulamos para $x$ y $f(x)$ con h=0.05

In [ ]:
x2,fx2 = Listas(0.5,0.05,30)
tabla2 = {'x': x2,'f(x)': fx2}
df2 = pd.DataFrame(tabla2)
df2.head(5)

x      f(x)
0  0.50  0.058463
1  0.55  0.078973
2  0.60  0.104151
3  0.65  0.134626
4  0.70  0.171071

Creamos y tabulamos para $x$ y $f(x)$ con h=0.025

In [ ]:
x3,fx3 = Listas(0.5,0.025,60)
tabla3 = {'x': x3,'f(x)': fx3}
df3 = pd.DataFrame(tabla3)
df3.head(5)

x      f(x)
0  0.500  0.058463
1  0.525  0.068172
2  0.550  0.078973
3  0.575  0.090941
4  0.600  0.104151

Implementamos la regla del trapecio

In [ ]:
def Regla_Trapecio(x,fx):
  h = x[1]-x[0]
  inte = (h/2)*(fx[0] + fx[-1] + (2*np.sum(fx[1:-1])))
  return inte

In [ ]:
def Meto(tap):
  resul=[]
  for i in range(len(tap)-1):
    resul.append(tap[i+1] + ((tap[i+1] - tap[i])/3))
  return resul

In [ ]:
def O6(tap):
  resul=[]
  for i in range(len(tap)-1):
    resul.append(tap[i+1] + ((tap[i+1] - tap[i])/15))  
  return resul

Aplicamos regla del trapecio

In [ ]:
tap = [Regla_Trapecio(x,fx),Regla_Trapecio(x2,fx2),Regla_Trapecio(x3,fx3)]

Finalmente obtenemos lo siguiente:

In [ ]:
H4 = Meto(tap)
H6 = O6(H4)
H4.insert(0,0)
H6.insert(0,0)
H6.insert(0,0)
tabla2 = {'h': [0.1,0.5,0.025],'n':[15,30,60], 'Trap O(h2)': tap, 'O(h4)': H4, 'O(h6)': H6}
df2 = pd.DataFrame(tabla2)
df2

h   n  Trap O(h2)     O(h4)     O(h6)
0  0.100  15    2.868982  0.000000  0.000000
1  0.500  30    2.861487  2.858989  0.000000
2  0.025  60    2.859613  2.858989  2.858989

Por lo tanto la estimación con orden del error 6, es:
$$
\int_{0.5}^{2} x^{3} \frac{\sqrt{3+2x^{2}}}{4} dx  ≈ 2.858989
$$


## [Regresar al ínidce](https://github.com/neto-riga/metodos_numericos_2/blob/main/portafolio.md#índice)